### DISCUSSION TITLE
Resetting camera when adding an actor.

### DISCUSSION THREAD

**Discussion by AndCotOli** [Kitware/trame#316] [Q&A] [answered]
Hi there!
Following [this example](https://github.com/pyvista/pyvista/blob/main/examples_trame/basic/file_viewer.py) I'm trying to build a web app that let's users view meshes.

The problem I'm facing is that, when the file changes, the camera doesn't reset and the mesh is out of place in the viewer. This is fixed by manually calling `ctrl.view_reset_camera` (with a button click in my case).

How can I update automatically the camera when a new file is loaded?

This is my state handling.
```python
@state.change("file_exchange")
def file_exchange(file_exchange, **kwargs):
    mesh, error = get_mesh_from_file_exchange(file_exchange)

    if error:
        return

    plotter.clear_actors() # pv.Plotter

    add_mesh_to_view(mesh, state)

    # Does nothing
    ctrl.view_reset_camera()
    ctrl.view_update()
```

And the code for adding a mesh to the plotter.
```python
def add_mesh_to_view(mesh, state):
    mesh.compute_normals(inplace=True)
    arrows = mesh.glyph(orient="Normals", tolerance=0.05, scale=False)
    plotter.add_mesh(arrows, name="normals", show_scalar_bar=False)
    plotter.actors["normals"].SetVisibility(state.normals_visibility)

    plotter.add_mesh(mesh, name="mesh")
```

Thanks for your help!

**Comment by jourdain** [comment]
That is a good one, but the key part is that PyVista probably uses the VtkLocalView (check the backend you choose).

With the VtkLocalView, it is the client that computes the bounds and resets the camera. So when you trigger it, the full scene is not yet available on the client side. Hence the mesh is out of place. 

There are 3 solutions. 

1. You can add an event listener on the VtkLocalView to trigger a reset camera once completed. (Probably not practical since PyVista is hiding that UI section. 
2. You can compute the camera position on the server (where you already have the data) and send that camera to the client. 
3. Force the rendering to happen remotely. 

For more detailed solution, someone will have to look at the PyVista code to know which method could be called. 

@alesgenova

**Comment by AndCotOli** [comment]
Pyvista is using `vtkRemoteLocalView` (which I suppose uses vtkLocalView when in local mode?), so forcing the rendering to happen remotely fixes the issue.

However, I'd like to still be able to render locally, so I'm thinking on going with option 3 since ) believe it's the most practical (server does less work and less information is transferred over WS).
I can't find any docs on how to do it though, do you have any ideas on how to do it?

**Comment by jourdain** [comment, marked as answer]
The PyVista viewer has a [`push_camera` method](https://github.com/pyvista/pyvista/blob/main/pyvista/trame/ui/base_viewer.py#L69). 

The number 3 is basically saying that if you are using VtkRemoteView, you are good to go... This means that you [force a given mode](https://github.com/pyvista/pyvista/blob/983e5775412dfe0c56aaff00b21a5fec59a6fe13/pyvista/trame/ui.py#L503-L507).

**Reply by AndCotOli** [reply]
Setting the camera position with `plotter.camera_position` and then using `view.push_camera` did the trick. Thanks!
